### PR TITLE
Minigui

### PR DESCRIPTION
**Pull request by tommadams**
Initial version of Minigui

### CODE REVIEW

File: minigui/graph.ts [hunk 1]
```diff
@@ -0,0 +1,190 @@
...
+
+    this.draw();
+  }
+
+  clear() {
+    this.points = [];
+    this.draw();
+  }
+
+  setPoint(x: number, y: number) {
```
Comment (amj) [inline minigui/graph.ts]: this probably needs a comment; looks like this is actually more like `addPoint`, semantically, but with some magic around not overscaling the opening moves?

File: minigui/gtp_socket.ts [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  send(cmd: string, callback?: CmdHandler) {
+    this.cmdQueue.push({cmd: cmd, callback: callback || null});
+    if (this.cmdQueue.length == 1) {
+      this.sendNext();
+    }
+  }
+
+  newSession() {
```
Comment (amj) [inline minigui/gtp_socket.ts]: a callout to where this session logic is handled in the python code will be helpful to noobs

File: minigui/gtp_socket.ts [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+  }
+
+  private cmdHandler(result: string) {
+    let {cmd, callback} = this.cmdQueue[0];
+    this.cmdQueue = this.cmdQueue.slice(1);
+
+    if (this.debugHandler) {
+      let trimmedCmd = cmd;
+      if (trimmedCmd.substr(0, 3) == 'mg_') {
+        trimmedCmd = cmd.substr(3);
```
Comment (amj) [inline minigui/gtp_socket.ts]: i think the behavior here might end up with folks thinking those are the GTP commands they need to type; i.e., someone who knows a bit about GTP will type "genmove b" and then wonder why no variations appear.  I totally didn't do this myself a few times.... of course not....

File: minigui/log.ts [hunk 4]
```diff
@@ -0,0 +1,71 @@
+import {getElement, querySelector} from './util';
+
+function escapeAnsiColors(raw: string) {
```
Comment (amj) [inline minigui/log.ts]: probably better to axe this and just send the non-colored version of the board over the wire, no?

File: minigui/minigui.ts [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+
+const N = board.BoardSize.Nine;
+
+namespace BoardState {
+  export type Move = board.Point | 'pass' | 'resign';
+}
+
+class BoardState {
+  stones: Array<board.Color>;
+  principleVariation = new Array<board.Move>();
```
Comment (amj) [inline minigui/minigui.ts]: I admire your commitment to princples ;)
https://en.wikipedia.org/wiki/Principal_variation_search
:rofl: (nit obvi)

File: minigui/minigui.ts [hunk 6]
```diff
@@ -0,0 +1,436 @@
...
+      }
+      if (result[2] == 'R') {
+        prettyResult += 'resignation';
+      } else {
+        prettyResult += result.substr(2) + ' points';
+      }
+
+      let img = document.createElement('img');
+      img.onload = () => { log.scroll(); };
+      img.src = `img/${1 + Math.floor(7 * Math.random())}.jpg`;
```
Comment (amj) [inline minigui/minigui.ts]: somewhere you should say what you expect the images to be named without making someone track this down.

File: minigui/serve.py [hunk 7]
```diff
@@ -0,0 +1,89 @@
...
+
+
+def std_bg_thread(stream):
+    global token
+
+    for line in p.__getattribute__(stream):
+        line = line.decode()
+        if line[-1] == "\n":
+            line = line[:-1]
+
```
Comment (amj) [inline minigui/serve.py]: ws

File: minigui/minigui.ts [hunk 8]
```diff
@@ -0,0 +1,436 @@
+import * as board from './board'
```
Comment (amj) [inline minigui/minigui.ts]: You need the license headers in the .ts files.

File: minigui/static/socketio/LICENSE [hunk 9]
```diff
@@ -0,0 +1,22 @@
+(The MIT License)
+
+Copyright (c) 2014-2018 Automattic <dev@cloudup.com>
```
Comment (amj) [inline minigui/static/socketio/LICENSE]: what's this?

File: minigui/static/require/LICENSE [hunk 10]
```diff
@@ -0,0 +1,45 @@
+Copyright jQuery Foundation and other contributors, https://jquery.org/
```
Comment (amj) [inline minigui/static/require/LICENSE]: also, what's this?

File: minigui/minigui.ts (hunk 5 above)
Comment (tommadams) [reply minigui/minigui.ts]: done

File: minigui/board.ts [hunk 11]
```diff
@@ -0,0 +1,539 @@
...
+const STAR_POINTS = {
+  [BoardSize.Nine]: [[2, 2], [2, 6], [6, 2], [6, 6], [4, 4]],
+  [BoardSize.Thirteen]: [[3, 3], [3, 9], [9, 3], [9, 9], [6, 6]],
+  [BoardSize.Nineteen]: [[3, 3], [3, 9], [3, 15],
+                         [9, 3], [9, 9], [9, 15],
+                         [15, 3], [15, 9], [15, 15]],
+};
+
+interface BoardOptions {
+  label?: string;
```
[the remark this answers is not in the document]
Comment (tommadams) [reply minigui/board.ts]: caption! that's a much better name

File: gtp_extensions.py [hunk 12]
```diff
@@ -118,11 +124,17 @@ class GoGuiMixin(gtp.Engine):
 
     def __init__(self, game_obj, name="gtp (python, gogui extensions)", version="0.1"):
         super().__init__(game_obj=game_obj, name=name, version=version)
+        self.session_id = ""
         self.known_commands += ["gogui-analyze_commands"]
 
     def cmd_gogui_analyze_commands(self, arguments):
         return "\n".join(["var/Most Read Variation/nextplay",
                           "var/Think a spell/spin",
+                          "var/Minigui genmove/mg_genmove",
```
[the remark this answers is not in the document]
Comment (tommadams) [reply gtp_extensions.py]: done

File: gtp_extensions.py [hunk 13]
```diff
@@ -161,6 +173,129 @@ def cmd_spin(self, arguments):
             print("gogui-gfx: VAR", moves_cols, file=sys.stderr, flush=True)
         return self.cmd_nextplay(None)
 
+    def _dbg_search(self, leaves):
```
[the remark this answers is not in the document]
Comment (tommadams) [reply gtp_extensions.py]: done.

File: gtp_extensions.py [hunk 14]
```diff
@@ -161,6 +173,129 @@ def cmd_spin(self, arguments):
...
+        return arguments
+
+
+    def cmd_mg_genmove(self, arguments):
+        game = self._game
+
+        start = time.time()
+        debug_interval = int(arguments)
+        current_readouts = game.root.N
+        last_dbg = -debug_interval
```
[the remark this answers is not in the document]
Comment (tommadams) [reply gtp_extensions.py]: added an explanatory comment

File: minigui/board.ts [hunk 15]
```diff
@@ -0,0 +1,539 @@
...
+    for (let i = 0; i < stones.length; ++i) {
+      this.stones[i] = stones[i];
+    }
+  }
+
+  getStone(p: Point) {
+    return this.stones[p.row * this.size + p.col];
+  }
+
+  setMark(p: Point, style: string) {
```
[the remark this answers is not in the document]
Comment (tommadams) [reply minigui/board.ts]: added a comment

File: minigui/board.ts [hunk 16]
```diff
@@ -0,0 +1,539 @@
...
+      let top = this.boardToCanvas(0, i);
+      let bottom = this.boardToCanvas(n - 1, i);
+      ctx.moveTo(0.5 + left.x, 0.5 + left.y);
+      ctx.lineTo(0.5 + right.x, 0.5 + right.y);
+      ctx.moveTo(0.5 + top.x, 0.5 + top.y);
+      ctx.lineTo(0.5 + bottom.x, 0.5 + bottom.y);
+    }
+    ctx.stroke();
+
+    // Draw start points.
```
[the remark this answers is not in the document]
Comment (tommadams) [reply minigui/board.ts]: done

File: minigui/graph.ts (hunk 1 above)
Comment (tommadams) [reply minigui/graph.ts]: added a comment and made the logic less obscure

File: minigui/gtp_socket.ts (hunk 2 above)
Comment (tommadams) [reply minigui/gtp_socket.ts]: done

File: minigui/gtp_socket.ts (hunk 3 above)
Comment (tommadams) [reply minigui/gtp_socket.ts]: done

File: minigui/serve.py (hunk 7 above)
Comment (tommadams) [reply minigui/serve.py]: done

File: minigui/minigui.ts (hunk 8 above)
Comment (tommadams) [reply minigui/minigui.ts]: done

**Comment by amj** [comment]
LGTM, thanks!

**Pull request merged**